File: Python3/Equal Row and Column Pairs.py
#1
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        n = len(grid)
        count = 0
        rows = {}

        for r in range(n):
            row = tuple(grid[r])
            rows[row]= 1 + rows.get(row, 0)
        
        for c in range(n):
            col = tuple(grid[i][c] for i in range(n))
            count += rows.get(col, 0)
            
        return count

#2 From Solution
class Solution:                                
    def equalPairs(self, grid: List[List[int]]) -> int:

        tpse = Counter(zip(*grid))                  # <-- determine the transpose
                                                    #     and hash the rows

        grid = Counter(map(tuple,grid))             # <-- hash the rows of grid. (Note the tuple-map, so
                                                    #     we can compare apples w/ apples in next step.)

        return  sum(tpse[t]*grid[t] for t in tpse)  # <-- compute the number of identical pairs

File: Golang/Equal Row and Column Pairs.go
func equalPairs(grid [][]int) int {
    rowMap := map[string]int{}
    colMap := map[string]int{}
    count := 0

    for i:=0; i<len(grid); i++{
        col := ""
        row := ""
        for j:=0; j<len(grid); j++{
            row += strconv.Itoa(grid[i][j]) + ","
            col += strconv.Itoa(grid[j][i]) + ","
        }
        rowMap[row]++
        colMap[col]++
    }

    for key, val := range(rowMap){
        count += val * colMap[key]
    }

    return count
}